Addded advanced exaplanation and examples for st.button

MathCatsAnd opened the pull request:
## 📚 Context
Buttons and what gets nested inside them continues to be a repeat theme in the forum, stack exchange, etc. I feel like a broken record with as many times as I've typed:
> Buttons aren't stateful. They return True on the page load resulting from their click, and then go back to False.

#675 

It would be helpful (and a huge time saver) to have a more robust "button repository" to point to.

## 🧠 Description of Changes
This PR proposes a new KB article all about buttons will a long list of examples showing working and non-working logical patterns. It begins with:
```
## TLDR

Buttons do not retain state. They return `True` on the page load resulting from
their click and then immediately go back to `False`. If you nest something
inside a button, then it will go away as soon as the user takes their next
action (because the page reloads and the button becomes `False`).
```

- [X] Menu/Navigation updated for users to access article
- [ ] Reorganize/re-categorize "Using Streamlit" section of KB to be more accessible/intuitive
- [ ] Host some of the code snippets for interactivity?
- [ ] Update the `st.button` page with link to new KB article as appropriate
- [X] I assume the site map will get updated by itself automatically

## 💥 Impact
Size:
- [X] Not small <!-- Everything else -->

## 🌐 References
This is cheeky, but largely correct: [the forum](https://discuss.streamlit.io/search?q=%40mathcatsand%20buttons)... 😝 
Closes #675

**Contribution License Agreement**
By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

MathCatsAnd committed:
Skeleton of KB Buttons Article

Beginning draft of KB Buttons article.

MathCatsAnd committed:
Added button code examples

Started filling in code examples of what to do and what not to do with buttons.

MathCatsAnd committed:
Added more button code examples

Continued filling in code examples of what to do and what not to do with buttons.

MathCatsAnd committed:
Finished button code examples, first pass

Filled in the rest of the code examples to complete a first pass at the button KB article.

MathCatsAnd commented:
@snehankekre I've marked this as a draft to get some feedback. (@jrieke do you want to be pinged too on these bigger ideas?) I'm not 100% in love with each and every example and I think a couple could be tweaked for clarity, but this is a first pass to get the idea "on paper" and get the ball rolling. 

I'd like to also reorganize the "Using Streamlit" section as the long list of articles doesn't feel very accessible. Maybe most people are getting to those articles through search, but it would be nice to have some grouping/categorization of the articles. I thought it might even be nice to create a separate "widgets" section if there are other widgets that could use a longer-winded explanation and plethora of toy examples beyond their basic API doc page.

<details><summary>Preview</summary>
<img src="https://github.com/streamlit/docs/assets/98661771/408dab2c-d4a0-41d9-bb6a-e176ce8c7315">
</details>

MathCatsAnd committed:
Updated per editor review

Updated for grammar and style per feedback from an editor

MathCatsAnd commented:
> @MathCatsAnd IIUC the recent feedback from KM has yet to incorporated here. Once that's done, I'll do a final review and merge 😄

Whoops, forgot to push. Done now.

snehankekre commented on content/kb/using-streamlit/index.md:
```diff
@@ -8,6 +8,7 @@ slug: /knowledge-base/using-streamlit
 Here are some frequently asked questions about using Streamlit. If you feel something important is missing that everyone needs to know, please [open an issue](https://github.com/streamlit/docs/issues) or [submit a pull request](https://github.com/streamlit/docs/pulls) and we'll be happy to review it!
 
 - [Sanity checks](/knowledge-base/using-streamlit/sanity-checks)
+- [Buttons, buttons, buttons!](/knowledge-base/using-streamlit/buttons)
```
Title here should ideally match that of the article.

snehankekre replied:
Edit slug to: /knowledge-base/using-streamlit/button-behavior-and-examples

snehankekre commented on content/menu.md:
```diff
@@ -473,6 +473,9 @@ site_menu:
   - category: Knowledge base / Using Streamlit / Batch elements and input widgets with st.form
     url: /knowledge-base/using-streamlit/batch-elements-input-widgets-form
     visible: false
+  - category: Knowledge base / Using Streamlit / Buttons, buttons, buttons!
```
Title here should ideally match that of the article.

snehankekre commented on content/kb/using-streamlit/buttons.md:
```diff
@@ -0,0 +1,412 @@
+---
+title: Button behavior and examples
+slug: /knowledge-base/using-streamlit/buttons
```
```suggestion
slug: /knowledge-base/using-streamlit/button-behavior-and-examples
```

Better this way for SEO.

snehankekre commented on content/menu.md:
```diff
@@ -473,6 +473,9 @@ site_menu:
     url: /knowledge-base/using-streamlit/batch-elements-input-widgets-form
     visible: false
+  - category: Knowledge base / Using Streamlit / Buttons, buttons, buttons!
+    url: /knowledge-base/using-streamlit/buttons
```
Same suggestion as above for the slug :)

snehankekre commented on content/kb/using-streamlit/buttons.md:
```diff
@@ -0,0 +1,412 @@
+
+## Use cases for buttons
+
+When code is conditioned on a button's value, it will execute once in
+response to the button being clicked and not again (until the button is clicked
+again).
```
Perhaps reword slightly or add some more text here? It's unclear why this sentence follows a heading that says use cases.

snehankekre commented on content/kb/using-streamlit/buttons.md:
```diff
@@ -0,0 +1,412 @@
+response to the button being clicked and not again (until the button is clicked
+again).
+
+### Nest inside of a button:
```
Nest what? 😅

snehankekre commented on content/kb/using-streamlit/buttons.md:
```diff
@@ -0,0 +1,412 @@
+- Once-per-click processes that saves data to session state, a file, or
+  a database.
+
+### Do not nest inside of a button:
```
Given there's no preamble before these two headings, I find it difficult to understand how these two headings are use cases. We need to flesh this section out a bit more for clarity.

snehankekre commented on content/kb/using-streamlit/buttons.md:
````diff
@@ -0,0 +1,412 @@
+If you want to give the user a quick button to check if an entry is valid,
+but not keep that check displayed as the user continues:
+
+```python
+import streamlit as st
+
+animal_shelter = ['cat', 'dog', 'rabbit', 'bird']
+
+animal = st.text_input('Type an animal')
+
+if st.button('Check availability') and animal.lower() in animal_shelter:
+    # This goes away if the user clicks the other button or changes the animal
+    st.write('We have that animal!')
+
+st.button('OK')
+```
````
We should expand on what this code does and how it relates to the preceding sentence. "In the above/below example, ...."

snehankekre commented on content/kb/using-streamlit/buttons.md:
```diff
@@ -0,0 +1,412 @@
+### Stateful button
+
+If you want a clicked button to continue to be `True`, create a value in
+`st.session_state` and use the button set it to `True` in a callback:
```
hyperlink to `/library/advanced-features/session-state`

sfc-gh-dmatthews committed:
Expanded on example explanations

Added more description on longer or complicated examples.

sfc-gh-dmatthews committed:
Updated advanced button slug

sfc-gh-dmatthews committed:
Punctuation

sfc-gh-dmatthews committed:
Grammar

sfc-gh-dmatthews commented on content/library/advanced-features/button-behavior-and-examples.md:
````diff
@@ -0,0 +1,418 @@
+    st.write(st.session_state.processed[option][0])
+```
+
+Astute observers may think, "This feels a little like caching." We are only saving results relative to one parameter, but the pattern could easily be expanded to save results relative to both parameters. In that sense, yes, it has some similarities to caching, but also some important differences. When you save results in `st.session_state`, the results are only available to the current user in th ...
````
Am I being too obtuse here?

sfc-gh-dmatthews committed:
Removed link for moved article

Removed link to button behavior from KB since it was moved to library/advanced features.
